Reword "object safety" as "dyn compatibility" (#1512)
rust-lang/reference: src/items/traits.md

}
```

> ## Trait-Object Safety <a id="object-safety"></a>

Reviewer: ```suggestion
## Dyn Compatibility
```

NB: I'd love the URL fragment (named anchor) to get changed to `dyn-compatibility` and have an HTTP redirect from `object-safety`. Although I haven't checked if the Reference is set up to support redirects. I know that the rustc-dev-guide is.

Reviewer: We should add a paragraph or footnote pointing out that this concept is "formerly known as object safety".

Reviewer: > NB: I'd love the URL fragment (named anchor) to get changed to dyn-compatibility and have an HTTP redirect from object-safety

This would be quite nice because we show this URL literally in rustc's diagnostics. As a result of that we currently have a mismatch of terminology (diagnostic text vs. URL). Of course, that's expected but I'd like this to be temporary only (namely until this PR hits stable (rustc's diagnostics usually link to the stable channel of the Reference)).

Reviewer 2: When renaming sections, you need to use a `<script>` tag to redirect the fragment, see https://github.com/rust-lang/reference/blob/c64e52a3d306eac0129f3ad6c6d8806ab99ae2e9/src/expressions/operator-expr.md?plain=1#L687-L699 as an example.

> An object-safe trait can be used in a `dyn Trait` type, and be the base trait of a [trait object].
> A trait is *object safe* if it has the following qualities (defined in [RFC 255]):

Reviewer: ```suggestion
A dyn-compatible trait can be used in a trait object type (`dyn Trait`), and be the base trait of a [trait object].
A trait is *dyn compatible* if it has the following qualities (defined in [RFC 255]):
```

Reviewer: Unrelated to the renaming: The dyn-compatibility rules have been updated over time (for example, "dyn-incompatible" associated *types* marked with `where Self: Sized`  no longer render the trait dyn-incompatible). So stating that the rules were defined in RFC 255 is misleading or at the very least outdated. That should probably be changed as part of a separate PR though.

> * All [supertraits] must also be trait-object safe.

Reviewer: ```suggestion
* All [supertraits] must also be dyn compatible.
```

# use std::rc::Rc;
# use std::sync::Arc;
# use std::pin::Pin;

> // Examples of trait-object safe methods.

Reviewer: ```suggestion
// Examples of dyn-compatible methods.
```

Though strictly speaking we haven't defined "dyn compatible" for methods, only for traits ("type mismatch"). We could be more explicit.

```

```rust,compile_fail

> // This trait is object-safe, but these methods cannot be dispatched on a trait object.

Reviewer: You should double-check if there are any other mentions of `/(non[\- _])?obj(ect)?[\- _](un)?safe(ty)?/i`. and update them accordingly.

```rust,compile_fail
# use std::rc::Rc;

> // Examples of non-object safe traits.

Reviewer: ```suggestion
 // Examples of dyn-incompatible traits.
```

\+ all other occurrences of this phrasing in this repo.

> trait NotDynCompatible {

Reviewer: ```suggestion
trait DynIncompatible {
```

\+ usage sites
